### ISSUE TITLE
MiniRocket for time series forecasting

### ISSUE THREAD

**Issue by SivaAndMe** [timeseriesAI/tsai#290]
Hi!
I would like to use **MiniRocket for forecasting metal price volatility**. I found out this [notebook](https://colab.research.google.com/gist/oguiza/d05d777973a74de2c9ccdc0e521c2837/scratchpad.ipynb#scrollTo=J82lmtym1-88) and I could train a forecasting model using MiniRocket architecture. However, I would like to use a SkLearn type API (because the above-mentioned way takes more time) as presented in this [notebook](https://colab.research.google.com/github/timeseriesAI/tsai/blob/master/tutorial_nbs/10_Time_Series_Classification_and_Regression_with_MiniRocket.ipynb#scrollTo=4FMoe-SEZGdt). In this notebook, I could find MiniRocket classifier & Minirocket regressor, but not Minirocket forecaster. @oguiza mentioned [here](https://github.com/timeseriesAI/tsai/issues/159#issuecomment-886901079) **that a regression problem can be converted to a forecasting problem by changing the target.** If that is possible, I can use the MiniRocketRegressor() for forecasting. (Can I ?). However, I am not sure how to do that. Can someone help me in this regard?

What is the difference between training the Minirocket models in the above two notebooks? 

Also, I will appreciate it if someone can comment on the performance of Rocket/Minirocket for time series forecasting. What are the best models available in tsai library for time series forecasting?

Thank you!

**Comment by oguiza** [comment]
Hi @SivaAndMe,
Good questions.
The first thing to bear in mind is that the sklearn-type API is recommended by the paper authors for up to 10k samples. I don't know how big is your dataset.
MiniRocketRegressor and MiniRocketForecaster would be the same thing. The model just calculates a float value. But it doesn't really care which value it is. In the case of a forecast, it'd be the next. In the case of regression, it's an independent value. So the answer is yes, you can try it. (A piece of advice Jeremy Howard from fastai gives in his courses is. "if you want to know if anything works try it!").

> What is the difference between training the Minirocket models in the above two notebooks?

MiniRocket is targeted at datasets of any size since the model is trained using batches. MiniRocketClassifier or MiniRocketRegressor works with all data at the same time. That's why there's a limit of 10k samples.

> Also, I will appreciate it if someone can comment on the performance of Rocket/Minirocket for time series forecasting. 

I'm not aware of any paper that has systematically reviewed MiniRocket's performance on time series forecasting datasets. But you can try it. It's very fast.

> What are the best models available in the tsai library for time series forecasting?

In my experience, Transformers (like TST or TSiT in tsai) tend to outperform other models in TSF tasks.

**Comment by SivaAndMe** [comment]
Thank you @oguiza!
My dataset size is ~6k. Hence, I'll work with MinirocketRegressor.

I have one more question. 

`lr_find()` suggests different learning rates in different runs. Is there any way to control this behavior? I have noticed this (different results with multiple runs) issue while training other models also (for example - InceptionTime). I encountered this problem mainly with `learner.fit_one_cycle()`. I remember trying to fix this issue with `torch.seed()` and other methods, but it didn't help. Is there any way to fix this?

Thanks again!

**Comment by vrodriguezf** [comment]
Make sure you seed everything. Here's a `seed_everything` function from [pytorch lightning](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/568a1e0a68fb8a2d8f2be7fae4f79c9068ca56fb/pytorch_lightning/utilities/seed.py):

```
def seed_everything(seed: Optional[int] = None, workers: bool = False) -> int:
    """Function that sets seed for pseudo-random number generators in: pytorch, numpy, python.random In addition,
    sets the following environment variables:

    - `PL_GLOBAL_SEED`: will be passed to spawned subprocesses (e.g. ddp_spawn backend).
    - `PL_SEED_WORKERS`: (optional) is set to 1 if ``workers=True``.

    Args:
        seed: the integer value seed for global random state in Lightning.
            If `None`, will read seed from `PL_GLOBAL_SEED` env variable
            or select it randomly.
        workers: if set to ``True``, will properly configure all dataloaders passed to the
            Trainer with a ``worker_init_fn``. If the user already provides such a function
            for their dataloaders, setting this argument will have no influence. See also:
            :func:`~pytorch_lightning.utilities.seed.pl_worker_init_function`.
    """
    max_seed_value = np.iinfo(np.uint32).max
    min_seed_value = np.iinfo(np.uint32).min

    if seed is None:
        env_seed = os.environ.get("PL_GLOBAL_SEED")
        if env_seed is None:
            seed = _select_seed_randomly(min_seed_value, max_seed_value)
            rank_zero_warn(f"No seed found, seed set to {seed}")
        else:
            try:
                seed = int(env_seed)
            except ValueError:
                seed = _select_seed_randomly(min_seed_value, max_seed_value)
                rank_zero_warn(f"Invalid seed found: {repr(env_seed)}, seed set to {seed}")
    elif not isinstance(seed, int):
        seed = int(seed)

    if not (min_seed_value <= seed <= max_seed_value):
        rank_zero_warn(f"{seed} is not in bounds, numpy accepts from {min_seed_value} to {max_seed_value}")
        seed = _select_seed_randomly(min_seed_value, max_seed_value)

    # using `log.info` instead of `rank_zero_info`,
    # so users can verify the seed is properly set in distributed training.
    log.info(f"Global seed set to {seed}")
    os.environ["PL_GLOBAL_SEED"] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

    os.environ["PL_SEED_WORKERS"] = f"{int(workers)}"

    return seed


def _select_seed_randomly(min_seed_value: int = 0, max_seed_value: int = 255) -> int:
    return random.randint(min_seed_value, max_seed_value)
```

**Comment by SivaAndMe** [comment]
Thank you @vrodriguezf  for your suggestion. But seed_everything() is not working. I tried with the code given in this [notebook](https://colab.research.google.com/gist/oguiza/d05d777973a74de2c9ccdc0e521c2837/scratchpad.ipynb#scrollTo=W0m27VPB0cM7). It is suggesting different LRs with different runs. Here's the code:
```
from tsai.all import *
p = !python  -V
print(f'python         : {p[0].split(" ")[1]}')
print(f'tsai           : {tsai.__version__}')
print(f'fastai         : {fastai.__version__}')
print(f'fastcore       : {fastcore.__version__}')
print(f'torch          : {torch.__version__}')
iscuda = torch.cuda.is_available()
print(f'device         : {device} ({torch.cuda.get_device_name(0)})' if iscuda else f'device         : {device}')
def seed_everything(seed: Optional[int] = None, workers: bool = False) -> int:
    Args:
        seed: the integer value seed for global random state in Lightning.
            If `None`, will read seed from `PL_GLOBAL_SEED` env variable
            or select it randomly.
        workers: if set to ``True``, will properly configure all dataloaders passed to the
            Trainer with a ``worker_init_fn``. If the user already provides such a function
            for their dataloaders, setting this argument will have no influence. See also:
            :func:`~pytorch_lightning.utilities.seed.pl_worker_init_function`.
    """
    # imprt 
    max_seed_value = np.iinfo(np.uint32).max
    min_seed_value = np.iinfo(np.uint32).min

    if seed is None:
        env_seed = os.environ.get("PL_GLOBAL_SEED")
        if env_seed is None:
            seed = _select_seed_randomly(min_seed_value, max_seed_value)
            rank_zero_warn(f"No seed found, seed set to {seed}")
        else:
            try:
                seed = int(env_seed)
            except ValueError:
                seed = _select_seed_randomly(min_seed_value, max_seed_value)
                rank_zero_warn(f"Invalid seed found: {repr(env_seed)}, seed set to {seed}")
    elif not isinstance(seed, int):
        seed = int(seed)

    if not (min_seed_value <= seed <= max_seed_value):
        rank_zero_warn(f"{seed} is not in bounds, numpy accepts from {min_seed_value} to {max_seed_value}")
        seed = _select_seed_randomly(min_seed_value, max_seed_value)

    # using `log.info` instead of `rank_zero_info`,
    # so users can verify the seed is properly set in distributed training.
    # log.info(f"Global seed set to {seed}")
    print(f"Global seed set to {seed}")
    os.environ["PL_GLOBAL_SEED"] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

    os.environ["PL_SEED_WORKERS"] = f"{int(workers)}"

    return seed


def _select_seed_randomly(min_seed_value: int = 0, max_seed_value: int = 255) -> int:
    return random.randint(min_seed_value, max_seed_value)
seed_everything(43)

feature_columns = ['a', 'b', 'c', 'my_target_var']
data = np.random.rand(1000, len(feature_columns),)
df = pd.DataFrame(data, columns=feature_columns)
df

window_length = 7
horizon = 1

X, y = SlidingWindow(window_length, horizon=horizon, get_x=feature_columns, get_y='my_target_var')(df)
splits = get_splits(y, valid_size=.3, stratify=True, random_state=43, shuffle=False)
X.shape, y.shape, splits

tfms  = [None, [TSForecasting()]] # breaks other methods
batch_tfms = TSStandardize(by_sample=True, by_var=True, verbose=True)
dls_inc = get_ts_dls(X, y, splits=splits, tfms=tfms, batch_tfms=batch_tfms, bs=128, device="cpu")
learn_inc = ts_learner(dls_inc, InceptionTime, metrics=[mae, mape, mse, rmse], cbs=ShowGraph())
# learn_inc.fit_one_cycle(5, 1e-2)
learn_inc.lr_find()
# SuggestedLRs(valley=0.0002290867705596611)
# SuggestedLRs(valley=0.0010000000474974513)
# SuggestedLRs(valley=0.0004786300996784121)

```

Thank you.

**Comment by oguiza** [comment]
You can try fastai's seed:

```bash
from fastai. torch_core import set_seed
set_seed(1234, reproducible=True) # replace 1234 by any integer you want use
```

Having said that, lr_find will give you an LR approximation. Even if you fix everything, you won't have a guarantee that you get the best LR possible. It'll be something close to it. The LRs you shared before are not too far from each other IMO.

**Comment by oguiza** [comment]
I will close this issue due to a lack of response. Feel free to reopen if necessary.

**Comment by SivaAndMe** [comment]
Thank you @oguiza . I have checked with fastai's seed also. It is still suggesting different LRs, but the reason you have given makes sense.
Is this problem (different results in different runs) supposed to happen while training a model also ? I have noticed that different runs with the same dataset, same model & same hyper parameters, gives different results. You can find the code [here](https://colab.research.google.com/drive/1Sb2rHArk7UwllyiMFZu38tir2IG8LsEh?authuser=1#scrollTo=OV907_kOcE8_). Can you please let me know how to overcome this problem?

Thank you!

**Comment by oguiza** [comment]
If you are really interested you can log a new issue since this issue is closed and you are bringing up something different from the original description. I'm not sure you'll be able to achieve completely deterministic results. 
Personally, I don't consider this a problem, but more a feature of neural networks. Even if you could have 100% deterministic results it doesn't mean you are selecting the best hyperparameters, since the results you get is just one within a distribution of possible outcomes.

**Comment by vrodriguezf** [comment]
> Personally, I don't consider this a problem, but more a feature of neural networks

This is true in the real world, but it can matter in challenges and in research papers in which you are asked for extreme reproducibility (this does not happen so often though xd). And I usually have the same problem as @SivaAndMe, I just cannot seed absolutely everything that is stochastic in Pytorch and I'm not sure why....not a tsai thing though

**Comment by oguiza** [comment]
I get your point. 
As I said before, please, feel free to log a new issue. It'll take me some time and I'm not 100% sure it can be worked out, but I'll take a look at it.
